sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java

```diff
@@ -110,16 +118,74 @@ public String apply(String input) {
                     }))
             .apply(View.<String>asSingleton());
 
-    PCollection<KV<TableDestination, TableRow>> inputInGlobalWindow =
+    PCollection<KV<DestinationT, TableRow>> inputInGlobalWindow =
         input.apply(
             "rewindowIntoGlobal",
-            Window.<KV<TableDestination, TableRow>>into(new GlobalWindows())
+            Window.<KV<DestinationT, TableRow>>into(new GlobalWindows())
                 .triggering(DefaultTrigger.of())
                 .discardingFiredPanes());
 
+    final TupleTag<KV<TableDestination, TableRow>> elementsOutputTag = new TupleTag<>();
+    final TupleTag<KV<String, String>> tableSchemasTag = new TupleTag<>();
+
+    List<PCollectionView<?>> resolveSideInputs = Lists.newArrayList();
+    if (dynamicDestinations.getSideInput() != null) {
+      resolveSideInputs.add(dynamicDestinations.getSideInput());
+    }
+    PCollectionTuple resolvedInputTuple = inputInGlobalWindow.apply(
+        "resolveInput",
+        ParDo.of(new DoFn<KV<DestinationT, TableRow>, KV<TableDestination, TableRow>>() {
+      DynamicDestinations<?, DestinationT> dynamicDestinationsCopy;
+
+      @StartBundle
+      public void startBundle(Context c) {
+        this.dynamicDestinationsCopy = SerializableUtils.clone(dynamicDestinations);
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) {
+        if (dynamicDestinationsCopy.getSideInput() != null) {
+          dynamicDestinationsCopy.setSideInputValue(
+              context.sideInput(dynamicDestinationsCopy.getSideInput()));
+        }
+        // Convert to a TableDestination and output.
+        TableDestination tableDestination =
+            dynamicDestinationsCopy.getTable(context.element().getKey());
+        TableReference tableReference = tableDestination.getTableReference();
+        if (Strings.isNullOrEmpty(tableReference.getProjectId())) {
+          tableReference.setProjectId(
+              context.getPipelineOptions().as(BigQueryOptions.class).getProject());
+          tableDestination =
+              new TableDestination(
+                  tableReference, tableDestination.getTableDescription());
+        }
+        context.output(elementsOutputTag, KV.of(tableDestination, context.element().getValue()));
+
+        // Match the schema, and output the mapping.
+        TableSchema tableSchema = dynamicDestinationsCopy.getSchema(context.element().getKey());
+        if (tableSchema != null) {
+          context.output(tableSchemasTag, KV.of(tableDestination.getTableSpec(),
```

Comment:
ends up being gross, since the map needs to be from tablespec->schema, as the following transforms no longer know about DestinationT (the reason I did it in one transform, as I didn't want to translate DestinationT->TableDestination multiple times). Maybe I'll look again at plumbing DestinationT into the following transforms, and get rid of the side-input map.